src/Ledger/TokenAlgebra/ValueSet.lagda

```diff
@@ -0,0 +1,222 @@
+  ⊕-lemma _ _ = lookup∘Fun⇒TotalMap-id
+
+  zeroFun : AssetId → Quantity
+  zeroFun = λ _ → 0
+
+  zeroMap : AssetId ⇒ Quantity
+  zeroMap = Fun⇒TotalMap zeroFun
+
+  ι : AssetId ⇒ Quantity
+  ι = Fun⇒TotalMap zeroFun
+
+  ι-lem0 :  ∀{a} → lookup ι a ≡ 0
+  ι-lem0 = ∈-rel→lookup-≡ ι (∈-map′ ⋁A)
+
+  ι-lem1 : ∀{a} → (a , lookup ι a) ∈ rel ι
+  ι-lem1 = lookup∈rel ι
+
+  ι-lem2 : ∀{a} → (a , 0) ∈ rel ι
+  ι-lem2 = ≡∈ ι-lem1 (Σ-≡,≡→≡ (≡.refl , ι-lem0))
+
+  ι-lem3 : ∀{a : AssetId} → lookup ι a ≡ 0
+  ι-lem3 = (left-unique-rel ι) ι-lem1 ι-lem2
+
+  _≋_ : stdlib.Rel (AssetId ⇒ Quantity) 0ℓ
+  u ≋ v = ∀{aa} → (lookup u) aa ≡ (lookup v) aa
+
+  ≋-isEquivalence : IsEquivalence {0ℓ} _≋_
+  ≋-isEquivalence = record { refl = ≡.refl ; sym = λ x → ≡.sym x ; trans = λ x y → ≡.trans x y }
+
+  ⊕-cong : Algebra.Congruent₂ _≋_ _⊕_
+  ⊕-cong {x}{y}{u}{v} x≋y u≋v {a} = begin
+    lookup (x ⊕ u) a         ≡⟨ ⊕-lemma x u ⟩
+    lookup x a + lookup u a  ≡⟨ cong (λ w → w + lookup u a) x≋y ⟩
+    lookup y a + lookup u a  ≡⟨ cong (λ w → lookup y a + w) u≋v ⟩
+    lookup y a + lookup v a  ≡⟨ ≡.sym (⊕-lemma y v) ⟩
+    lookup (y ⊕ v) a         ∎
+
+  ⊕-comm : Algebra.Commutative _≋_ _⊕_
+  ⊕-comm u v {aa} = begin
+    lookup (u ⊕ v) aa          ≡⟨ ⊕-lemma u v ⟩
+    lookup u aa + lookup v aa  ≡⟨ +-comm (lookup u aa) (lookup v aa) ⟩
+    lookup v aa + lookup u aa  ≡⟨ ≡.sym (⊕-lemma v u) ⟩
+    lookup (v ⊕ u) aa          ∎
+
+  ⊕-assoc : Algebra.Associative _≋_ _⊕_
+  ⊕-assoc x y z {a} = begin
+    lookup ((x ⊕ y) ⊕ z) a                  ≡⟨ ⊕-lemma (x ⊕ y) z ⟩
+    lookup (x ⊕ y) a + lookup z a           ≡⟨ cong (λ w → w + lookup z a) (⊕-lemma x y) ⟩
+    lookup x a + lookup y a + lookup z a    ≡⟨ +-assoc (lookup x a) (lookup y a) (lookup z a) ⟩
+    lookup x a + (lookup y a + lookup z a)  ≡⟨ cong (λ w → lookup x a + w) (≡.sym (⊕-lemma y z)) ⟩
+    lookup x a + lookup (y ⊕ z) a           ≡⟨ ≡.sym (⊕-lemma x (y ⊕ z)) ⟩
+    lookup (x ⊕ (y ⊕ z)) a                  ∎
+
+  ι-identity : Algebra.Identity _≋_ ι _⊕_
+  ι-identity .proj₁ tm {aa} = begin
+    lookup (ι ⊕ tm) aa          ≡⟨ ⊕-lemma ι tm ⟩
+    lookup ι aa + lookup tm aa  ≡⟨ cong (λ x → x + lookup tm aa) ι-lem3 ⟩
+    lookup tm aa                ∎
+  ι-identity .proj₂ tm {aa} = begin
+    lookup (tm ⊕ ι) aa          ≡⟨ ⊕-lemma tm ι ⟩
+    lookup tm aa + lookup ι aa  ≡⟨ cong (λ x → lookup tm aa + x) ι-lem3 ⟩
+    lookup tm aa + 0            ≡⟨ +-identityʳ (lookup tm aa) ⟩
+    lookup tm aa                ∎
+
+  open IsSemigroup
+  open IsMagma
+  isSemigrp : IsSemigroup _≋_ _⊕_
+  isSemigrp .isMagma .isEquivalence        = ≋-isEquivalence
+  isSemigrp .isMagma .∙-cong {u}{v}{x}{y}  = ⊕-cong {u}{v}{x}{y} -- (broken inference)
+  isSemigrp .assoc                         = ⊕-assoc
+
+  open IsMonoid
+  ≋-⊕-ι-isMonoid : IsMonoid _≋_ _⊕_ ι
+  ≋-⊕-ι-isMonoid .isSemigroup  = isSemigrp
+  ≋-⊕-ι-isMonoid .identity     = ι-identity
+\end{code}
+
+We are now in a position to define the commutative monoid.
```

Comment:
TODO: possibly add some more explanatory prose about the `Value-TokenAlgebra` example.